### PR TITLE
Refactor petsc compatibility

### PR DESCRIPTION
**Pull request by drwells**
This PR continues the work in #2754 by consolidating more version checks into one place. Additionally, several of the destructors used to contain `AssertThrow` checks; I removed them (destructors should not throw).

### CODE REVIEW

File: include/deal.II/lac/petsc_compatibility.h [hunk 1]
```diff
@@ -33,6 +36,12 @@ DEAL_II_NAMESPACE_OPEN
 
 namespace PETScWrappers
 {
+#if DEAL_II_PETSC_VERSION_LT(3,2,0)
+  typedef PetscTruth PetscBooleanType;
```
Comment (bangerth) [inline include/deal.II/lac/petsc_compatibility.h]: Aw, how old is this? Do we really need to support versions that are this old?

File: include/deal.II/lac/petsc_compatibility.h [hunk 2]
```diff
@@ -42,9 +51,116 @@ namespace PETScWrappers
...
+   *
+   * @warning The argument option_value is ignored in versions of PETSc
+   * before 3.0.0 since the corresponding function did not take this argument.
+   */
+  inline void set_matrix_option (Mat &matrix,
+                                 const MatOption option_name,
+                                 const PetscBooleanType option_value = PETSC_FALSE)
+  {
+#if DEAL_II_PETSC_VERSION_LT(3,0,0)
+    const int ierr = MatSetOption (matrix, option_name);
```
Comment (bangerth) [inline include/deal.II/lac/petsc_compatibility.h]: same here. We could make the code so much simpler if we got rid of backward compatibility with really old versions of PETSc.

**Comment by drwells** [comment]
It looks like (unless I am misreading something, see the link) Ubuntu 12.04 still ships 3.1, so I think that we should support at least that. I have no idea what ancient versions of PETSc people have on their clusters in practice. The CMake scripts check for PETSc 3.0 or newer right now.

We also have `#ifdefs` for 2.3. We can probably get rid of those :)

Fun fact: apparently PETSc 3.4 did not change anything, but every other release past (and including) 2.3 prompted an `#ifdef` in the library.

http://packages.ubuntu.com/search?keywords=petsc&searchon=names&suite=precise&section=all

**Comment by bangerth** [comment]
I see. Ubuntu 12.04 is more than 4 years old now, but it may still be on some people's laptops. I'm ok with supporting it if you think that's useful. PETSc 2.3 may be a bit old by now, though :-)

Re fun fact: Yes, I'm so not surprised :-(

**Comment by tamiko** [comment]
Acked. Very nice work!

**Comment by tamiko** [comment]
@drwells Have you had the chance to test your patchset against different versions of the PETSc library?
I can test the following versions:

```
tamiko@jackdaw ~ % eix -I petsc
[I] sci-mathematics/petsc [1]
     Available versions:  (~)3.4.2 (~)3.5.4 (~)3.6.4 (~)3.7.2 {X afterimage boost complex-scalars cxx debug doc fftw fortran hdf5 hypre mpi mumps scotch sparse superlu threads}
     Installed versions:  3.7.2(11:43:59 AM 07/09/2016)(X cxx doc fortran hdf5 hypre mpi mumps scotch sparse -afterimage -boost -complex-scalars -debug -fftw -superlu -threads)
     Homepage:            http://www.mcs.anl.gov/petsc/
     Description:         Portable, Extensible Toolkit for Scientific Computation
```

**Comment by drwells** [comment]
@tamiko I have not (I have just run the most recent). It is pretty easy for me to check, though (I'll try 3.6 in a moment and work back from that).

What versions of PETSc do we currently test the library with?

**Comment by tamiko** [comment]
@drwells 3.6.3 exclusively, if I'm not mistaken.

**Comment by bangerth** [comment]
I test 3.4.2

**Comment by drwells** [comment]
It looks like we can compile things with 3.3 and newer: the tests pass and there are no compilation warnings. Our CMake scripts cannot handle 3.2 since the directory layout changed.

**Comment by bangerth** [comment]
Ah, interesting. This would mean that we may as well remove all `#ifdef`s that test for anything older than 3.3 anyway. Or would you alternately want to extend the cmake scripts? When did our cmake configuration forget to deal with older PETSc versions?

**Comment by drwells** [comment]
I am not sure that we ever *de facto* supported these versions of PETSc. `FindPETSC.cmake` has not been changed much since 2014.

I would not mind going back and seeing what we need to do to support 3-1 - 3.3, but I think it is beyond the scope of the current PR (and I won't have time to get around to it for another week).

**Comment by drwells** [comment]
I checked the logs and it looks like I ignored an error message (no 64 bit hypre available) that I should not have when compiling 3.2: all of the tests passed when I gave PETSc a more reasonable set of compilation flags. My mistake.

3.2 works but raises two warnings which I fixed.

**Comment by drwells** [comment]
We can also compile the library with 3.1 but some of the tests fail for difficult to understand reasons. `tests.h` also needed a PETSc version check.

**Comment by bangerth** [comment]
This all works for me. Feel free to merge within a day if nobody else speaks up.

**Comment by tamiko** [comment]
Acked.

**Pull request merged**